test: kernel: Exclude broken LTO builds

rettichschnidi opened the pull request:
This disables the LTO builds for platforms which are affected by build errors.

While not a solution for the underlying issue, those changes will allow unrelated PRs with changes in the kernel code to pass CI and get merged.

Hint: https://github.com/zephyrproject-rtos/zephyr/pull/70426 tries to resolve the same issue

rettichschnidi committed:
tests: kernel: common: lto: Exclude builds with warning

Trying to enable LTO when CODE_DATA_RELOCATION is already enabled
produces a warning.

This commit prevents twister from building such combinations for
kernel.common.lto.

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi committed:
tests: kernel: interrupt: lto: Exclude builds with warning

Trying to enable LTO when CODE_DATA_RELOCATION is already enabled
produces a warning.

This commit prevents twister from building such combinations for
arch.shared_interrupt.lto.

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

danieldegrasse commented:
If we want to use this solution, we should disable builds based on if `CONFIG_CODE_RELOCATION` is selected- this is the actual underlying issue that is affecting these platforms, and could affect other ones as well

rettichschnidi commented:
> If we want to use this solution, we should disable builds based on if `CONFIG_CODE_RELOCATION` is selected- this is the actual underlying issue that is affecting these platforms, and could affect other ones as well

Just updated the PR. I guess you meant `CONFIG_CODE_DATA_RELOCATION`.

tejlmand reviewed (changes requested):
I think it's better to prevent LTO with Code relocation enabled.
At least until it has been ensured that those two features generally works together.

Thus instead of the changes proposed in this PR, then make an extra dependency here:
https://github.com/zephyrproject-rtos/zephyr/blob/1408d1e5b8cefcb0f4d52c1ffb41849f70826f39/Kconfig.zephyr#L430-L432

like:
```
depends on (!(GEN_ISR_TABLES || GEN_IRQ_VECTOR_TABLE) || ISR_TABLES_LOCAL_DECLARATION) && !NATIVE_LIBRARY && !CODE_DATA_RELOCATION
```

@rakons have you done any verification related to LTO and the code relocation feature ?

rettichschnidi committed:
kconfig: lto: Disable when data relocation is enabled

As per issue #69730, building arch.shared_interrupt.lto and
kernel.common.lto for mimxrt685_evk or mimxrt595_evk/mimxrt595s/cm33 is
currently broken due to their usage of CONFIG_CODE_DATA_RELOCATION.

This commit disables LTO when CODE_DATA_RELOCATION is enabled, allowing
PRs with changes in the kernel code to pass the CI check.

On the down side, builds affected by this change produce a new warning:

> warning: LTO (defined at Kconfig.zephyr:430) was assigned the value
> 'y' but got the value 'n'. Check these unsatisfied dependencies:
> (!CODE_DATA_RELOCATION) (=n)

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi commented:
> I think it's better to prevent LTO with Code relocation enabled.
> At least until it has been ensured that those two features generally works together.

I created a (draft) PR with this idea: https://github.com/zephyrproject-rtos/zephyr/pull/70426

While it also resolves (works around) my issue of no longer being able to pass CI with unrelated code changes, I am not really a fan of it as it produce a new warning:

> warning: LTO (defined at Kconfig.zephyr:430) was assigned the value 'y' but got the value 'n'. Check these unsatisfied dependencies: (!CODE_DATA_RELOCATION) (=n)

danieldegrasse commented:
> I think it's better to prevent LTO with Code relocation enabled.
At least until it has been ensured that those two features generally works together.

@tejlmand Do you think it would be possible to exclude the files relocated with code relocation using the `prohibit_lto` compiler property, like it is done here?: https://github.com/zephyrproject-rtos/zephyr/blob/main/CMakeLists.txt#L815

I am not sure if we can, since the code relocation feature uses generator expressions heavily. I tried to implement this to fix #69730 but did not get very far.

tejlmand commented:
> > I think it's better to prevent LTO with Code relocation enabled.
> > At least until it has been ensured that those two features generally works together.
> 
> @tejlmand Do you think it would be possible to exclude the files relocated with code relocation using the `prohibit_lto` compiler property, like it is done here?: https://github.com/zephyrproject-rtos/zephyr/blob/main/CMakeLists.txt#L815
> 
> I am not sure if we can, since the code relocation feature uses generator expressions heavily. I tried to implement this to fix #69730 but did not get very far.

I believe it should be possible.
I'm occupied with HWMv2 atm and won't have time to look into the details.

Implementing it should not take long but testing / verifying that things indeed work as expected is where I believe most time will need to be spend.

tejlmand commented:
@rettichschnidi I believe a combination of this PR (which filters out tests that enables data relocation) with the proposal in #70426 to avoid enabling LTO together with relocation will be the optimal solution to this problem.

andyross commented on Kconfig.zephyr:
```diff
@@ -429,7 +429,9 @@ endchoice
-	depends on (!(GEN_ISR_TABLES || GEN_IRQ_VECTOR_TABLE) || ISR_TABLES_LOCAL_DECLARATION) && !NATIVE_LIBRARY
+	depends on (!(GEN_ISR_TABLES || GEN_IRQ_VECTOR_TABLE) || ISR_TABLES_LOCAL_DECLARATION) \
+	  && !NATIVE_LIBRARY \
+	  && !CODE_DATA_RELOCATION
```
You can split the outer "&&"'s into separate "depends on" lines for readability.   Also "depends on !X" relationships are IMHO better expressed as a BUILD_ASSERT() in the C code somewhere, because there you document it as a straightforward error ("/* LTO not compatible with NATIVE_LIBRARY */") and not an inscrutable kconfig dependency failure.

rettichschnidi replied:
Will improve, then re-request a review from you.

rettichschnidi replied:
I incorporated the suggestion regarding the multiple "depends on" lines.

Did not do the BUILD_ASSERTS() suggestion, because I'd rather have Kconfig complaining about/enforcing things early on, not delaying the error until the build is happening.
